File: sdk/gsusb-csharp/examples/GsUsb.Examples/Examples/04_multi_std_asyncio.cs
using GsUsb;

namespace GsUsb.Examples;

internal static class Example04MultiStdAsyncio
{
    public static int Run(string[] args)
    {
        var options = CommonOptions.Parse(args, "4-channel standard CAN async example");
        var useFd = !options.Classic;
        var channels = new[] { 0, 1, 2, 3 };
        Console.WriteLine($"Pair topology: {DemoCommon.BuildPairTopologyText(channels)}");
        return DemoCommon.RunMultiSummaryDemo(
            $"Running 4-channel async demo for {options.DurationSeconds:F6}s...",
            "4-channel async summary",
            channels,
            options,
            WorkerStyle.Async,
            channelIndex => new GsUsbBus(DemoCommon.MakeBusOptions(options, channelIndex, useFd)),
            (channelIndex, counter) => new CanMessage
            {
                ArbitrationId = 0x300u + (uint)channelIndex,
                IsFd = useFd,
                BitrateSwitch = useFd,
                Data = DemoCommon.MakePayload([(byte)channelIndex, (byte)(counter & 0xFF), (byte)((counter >> 8) & 0xFF), 0x31, 0x41, 0x59, 0x26, 0x53], useFd ? 12 : 8, 0x58),
            });
    }
}

File: sdk/gsusb-cpp/examples/04_multi_std_asyncio.cpp
#include "demo_common.hpp"

#include <iostream>

int main(int argc, char** argv) {
    try {
        const auto options = demo::parse_common_options(argc, argv, "4-Channel Standard CAN Async Example");
        const bool use_fd = !options.classic;
        const std::vector<int> channels = {0, 1, 2, 3};
        std::cout << "Pair topology: " << demo::build_pair_topology_text(channels) << "\n";
        return demo::run_multi_summary_demo(
            "Running 4-channel async demo for " + std::to_string(options.duration_s) + "s...",
            "4-channel async summary",
            channels,
            options,
            demo::WorkerStyle::Async,
            [&](int channel_index) {
                return std::make_unique<gsusb::GsUsbBus>(demo::make_bus_options(options, channel_index, use_fd));
            },
            [&](int channel_index, std::uint64_t counter) {
                gsusb::CanMessage msg;
                msg.arbitration_id = 0x300u + static_cast<std::uint32_t>(channel_index);
                msg.is_fd = use_fd;
                msg.bitrate_switch = use_fd;
                msg.data = demo::make_payload({static_cast<std::uint8_t>(channel_index), static_cast<std::uint8_t>(counter & 0xFF), static_cast<std::uint8_t>((counter >> 8) & 0xFF), 0x31, 0x41, 0x59, 0x26, 0x53}, use_fd ? 12u : 8u, 0x58);
                return msg;
            }
        );
    } catch (const std::exception& ex) {
        std::cerr << ex.what() << "\n";
        return 1;
    }
}